builder/builder.go

```diff
@@ -0,0 +1,77 @@
+package builder
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/docker/docker/daemon"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/runconfig"
+)
+
+type ImageID string
+
+type Builder interface {
+	// TODO: make this return a reference and remove ImageID
+	Build(Docker, Context) ImageID
+}
+
+type ImageCache interface {
+	GetCachedImage(ImageID, *runconfig.Config) (ImageID, error)
+	Cache(ImageID, *runconfig.Config) error
+}
+
+type Docker interface {
+	// use digest reference instead of name
+	LookupImage(name string) (*image.Image, error)
+	Pull(name string) (*image.Image, error)
+
+	// move daemon.Container to its own package
+	Container(id string) (*daemon.Container, error)
+	Create(*runconfig.Config, *runconfig.HostConfig) (*daemon.Container, error)
+	Remove(id string, cfg *daemon.ContainerRmConfig) error
+	Commit(*daemon.Container, *daemon.ContainerCommitConfig) (*image.Image, error)
+	Copy(c *daemon.Container, destPath string, srcInfo ContextEntry, decompress bool) error
+}
+
+type Context interface {
+	io.Closer
+	Get(path string) (ContextEntry, error)
+	Open(path string) (io.ReadCloser, error)
+	Walk(filepath.WalkFunc) error
+}
+
+type ModifiableContext interface {
+	Context
+	Create(path string) (io.WriteCloser, error)
+	Remove(path string) error
+}
+
+type ContextEntry interface {
+	os.FileInfo
+	Path() string
+	Open() (io.ReadCloser, error)
+}
+
+type FileInfo struct {
+	os.FileInfo
+	FilePath string
+}
+
+func (fi FileInfo) Path() string {
+	return fi.FilePath
+}
+
+func (fi FileInfo) Open() (io.ReadCloser, error) {
+	return os.Open(fi.FilePath)
```

Comment:
Having this as `os.Open`, rather than going through the context breaks the interface of `Context`. I'd recommend removing the requirement for the `Open` method on `ContextEntry`.